app/script/cryptography/CryptographyRepository.coffee

```diff
@@ -393,25 +177,60 @@ class z.cryptography.CryptographyRepository
   ###
   encrypt_generic_message: (user_client_map, generic_message, payload = @_construct_payload @current_client().id) =>
-    @get_sessions user_client_map
-    .then (cryptobox_session_map) =>
-      return @_add_payload_recipients payload, generic_message, cryptobox_session_map
-
-  ###
-  Add the encrypted message for recipients to the payload message.
+    future_cipher_payloads = []
 
-  @private
-  @param payload [Object] Payload to add encrypted message for recipients to
-  @param cryptobox_sessions [Array<cryptobox.CryptoboxSession>] Sessions for all the recipients of message
-  @param generic_message [z.proto.GenericMessage] ProtoBuffer message to be send
-  @return [Object] Payload to send to backend
-  ###
-  _add_payload_recipients: (payload, generic_message, cryptobox_session_map) ->
-    for user_id, client_session_map of cryptobox_session_map
+    for user_id, client_ids of user_client_map
       payload.recipients[user_id] ?= {}
-      for client_id, cryptobox_session of client_session_map
-        payload.recipients[user_id][client_id] = @_encrypt_payload_for_session cryptobox_session, generic_message
-    return payload
+      client_ids.forEach (client_id) =>
+        session_id = @_construct_session_id user_id, client_id
+        future_cipher_payloads.push @_encrypt_payload_for_session session_id, generic_message
+
+    @logger.log "Encrypting message of type '#{generic_message.content}' for '#{Object.keys(payload.recipients).length}' users.", payload.recipients
+
+    z.util.PromiseUtil.execute_all future_cipher_payloads
+    .then (future_cipher_payloads) =>
+      user_client_map_for_missing_sessions = {}
+
+      future_cipher_payloads.results.forEach (result) ->
+        if result.encrypted
+          payload.recipients[result.user_id][result.client_id] = result.encrypted
+        else
+          user_client_map_for_missing_sessions[result.user_id] ?= []
+          user_client_map_for_missing_sessions[result.user_id].push result.client_id
+
+      return @_encrypt_generic_message_for_new_sessions user_client_map_for_missing_sessions, generic_message
+    .then (additional_cipher_payloads) ->
+      additional_cipher_payloads.forEach (result) ->
+        payload.recipients[result.user_id] ?= {}
+        payload.recipients[result.user_id][result.client_id] = result.encrypted
+      return payload
+
+  _encrypt_generic_message_for_new_sessions: (user_client_map_for_missing_sessions, generic_message) =>
+    return new Promise (resolve) =>
+      if Object.keys(user_client_map_for_missing_sessions).length
+        @get_users_pre_keys user_client_map_for_missing_sessions
+        .then (user_pre_key_map) =>
+          @logger.info "Fetched pre-keys for '#{Object.keys(user_pre_key_map).length}' users.", user_pre_key_map
+
+          future_sessions = []
+
+          for user_id of user_pre_key_map
+            for client_id of user_pre_key_map[user_id]
+              remote_pre_key = user_pre_key_map[user_id][client_id]
```

Comment:
```
@logger.log "Initializing session with Client ID '#{client_id}' from User ID '#{user_id}' with remote PreKey ID '#{remote_pre_key.id}'."
  session_id = @_construct_session_id user_id, client_id
  decoded_prekey_bundle_buffer = bazinga64.Decoder.fromBase64(remote_pre_key.key).asBytes.buffer
  return @cryptobox.session_from_prekey session_id, decoded_prekey_bundle_buffer
```

thats the same code used in line 115. might be good to create a function that does this